### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) and Post-Link Optimization (PLO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [apache/iggy#990] [Ideas]
Hi!

Recently I checked many optimizations like PGO and PLO (mostly with [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) improvements on multiple projects including many databases - the results are available [here](https://github.com/zamazan4ik/awesome-pgo/). According to the tests, these optimizations can help to achieve better performance in many cases. I think trying to optimize Iggy's performance with PGO and PLO can be an interesting idea.

I can suggest the following action points:

* Perform PGO on the database in more scenarios. If it shows improvements - add a note to the documentation about possible improvements in the project performance with PGO.
* Providing an easier way (e.g. a build option) to build scripts with PGO can be helpful for the end-users and maintainers since they will be able to optimize `iggy-server` according to their workloads.
* Testing Post-Link Optimization techniques (like [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) would be interesting too (Clang and Rustc already use BOLT as an addition to PGO) but I recommend starting from the usual PGO.

Here are some examples of how PGO optimization is integrated into other projects:

* Rustc: a CI [script](https://github.com/rust-lang/rust/tree/master/src/tools/opt-dist) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback"
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)

Unfortunately, for now, building the project with LTO and PGO enabled at the same time fails due to a bug in the compiler: https://github.com/rust-lang/rust/issues/115344#issuecomment-2150416060 . I suggest waiting for the fix in the upstream since disabling LTO is a worse option (IMHO).

I would be happy to answer your questions about all the optimizations above.

**Comment by hubcio** [comment]
Hello, I tested PGO with BOLT like 6 months ago, we got ~15% performance improvement. We have this issue:
https://github.com/iggy-rs/iggy/issues/200

I have branch somewhere with kind-of basic flow (build with pgo, run performance tests, build again, build with bolt, etc.)

For now, we simply don't have enough time to pursuit this :) I know this is really nice low hanging fruit and we will probably do it in future.

**Reply by zamazan4ik** [reply]
Wow, such a nice improvement - it's great. I hope one day you will find enough time for PGO integration into the main iggy-rs development branch.

**Comment by hubcio** [comment]
Also, thank you! I will ping you if I'll have more questions - did you already join our discord?

**Reply by zamazan4ik** [reply]
I didn't know about it before - just joined. Thanks!
